### PR TITLE
LPS-127012, LPS-129571 Create BalloonEditor component and sample portlet

### PR DESCRIPTION
**Pull request by julien**
These are just the base files for our `BalloonEditor` component, which
contains a React component using CKEditor's `balloontoolbar` plugin with the
minimum required configuration, a Java class and its corresponding JSP.

Additionally this PR contains a sample portlet in the `frontend-editor-ckeditor-sample-web`
module: the purpose of this portlet, is to be able to test our changes
during the implementation of the `BalloonEditor` component.

To add this portlet, edit the main page (or either create another widget page)
and look for CKEditor Sample in the widgets section.

For the moment this is what it looks like:

frontend-editor-ckeditor-sample-web

**Notes**
- This PR was originally sent and reviewed [here](https://github.com/liferay-frontend/liferay-portal/pull/932),
but got reverted [here](https://github.com/liferay-frontend/liferay-portal/pull/964) to fix [LPS-130399](https.//issues.liferay.com/browse/LPS-130399)

- For the moment I think this should be a draft PR to avoid it from being 
forwarded before everyone has had a change to have a look at it

### CODE REVIEW

**Review by ambrinchaudhary — APPROVED** [review]

**Comment by edalgrin** [comment]
Just started reviewing :)

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (jbalsas) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:48]: I'm wondering if this should be `CKEDITOR.env.cssClass += ' lfr_balloon_editor'` or `CKEDITOR.env.cssClass = `${CKEDITOR.env.cssClass} lfr_balloon_editor`.

Also I think our convention is to use dashes instead of underscores for our CSS classes. I'd consult an expert here, but we might want to make this `lfr-balloon-editor`. I know CKEditor's approach is different, but since this is our own code, I think it makes sense it follows our own rules.

Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:48]: About the `CKEDITOR.env.cssClass`, you're right: it might be dangerous to replace it with our own class ( @ambrinchaudhary asked me the same thing via Slack) 

I'll got with dashes in the class name, and use our conventions as much as possible.

**Comment by ambrinchaudhary** [comment]
Hey @liferay-frontend! I just marked this pull as ready for review, as I need it for rebasing another [pull](https://github.com/liferay-frontend/liferay-portal/pull/960). So any comment would be more than appreciated before forwarding it. Thanks all!

File: modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json:7]: This should be named `@liferay/frontend-ckeditor-sample-web` to guard against [dependency confusion attacks](https://github.com/liferay/liferay-frontend-projects/blob/master/guidelines/general/security/dependency_confusion.md).

We [would like to be able to lint for this](https://github.com/liferay/liferay-frontend-projects/issues/469), but we can't do that until we've renamed all of our existing packages ([LPS-129670](https://issues.liferay.com/browse/LPS-129670) (or we implement a whitelist). cc @jbalsas, @dsanz about prioritization of that work. (My feeling: this will be soul-destroying grunt work, but I'll feel nervous until we do it... and if we don't want to do it, we should hack in the whitelist so that we can deploy a lint soon.)

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:37]: Just to note a potential point of fragility here. _If_ this component ever re-rerenders, and _if_ it doesn't reliably trigger `onBeforeLoad` and `onDestroy` in perfect matching pairs, it is possible that the value that gets restored on line 51 will not be the one that got saved on line 46...

And also, that looks like a global variable... what's going to happen if there are multiple editors on the page fighting to save/restore the "initial" class? 😱 (Maybe this was discussed previously, so please forgive me if I missed that.)

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:70]: Using a `Set`.

god-bless-you-obama

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:78]: The `Set` is good for `O(1)` insertion, deletion, membership testing, but you are using it here for iteration `O(n)` (via `forEach`) and then you have a `path.contains()` which is itself a linear scan, so that's quadratic.

I want my Obama gif back!

What kind of object is `path`?

**Comment by wincent** [comment]
> So any comment would be more than appreciated before forwarding it. Thanks all!

Given that we had to revert this once, I think we should wait for the release to be 100% confirmed before we merged. It's just not worth the risk of getting it wrong a second time (because it would get us onto _la liste de merde_, as @julien would say)... 😂

File: modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json
Comment (dsanz) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json:7]: Having a white list may be read as "we want these to not have a namespace", when perhaps it should be taken as "that's the amount of pending work". Could we differentiate those in some way? And, are there chances that, once we rename all of our packages, whitelist becomes empty?

**Comment by julien** [comment]
I need to (fully) read and reply to your other comments, but I wanted to reply on this one first.

I totally agree that we could wait before merging this in right now, because

1. It's not finished 
2. @ambrinchaudhary has access to my liferay-portal fork so she can fetch, push or even delete the whole thing. 
3. I've skimmed through the comments @wincent left and there is stuff that needs to be fixed

I think there's no problem having this "sitting" here for a while and being a bit more patient: I definitely don't want to be on that list :joy:

File: modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json
Comment (wincent) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json:7]: > Could we differentiate those in some way? And, are there chances that, once we rename all of our packages, whitelist becomes empty?

Yeah, the "white list" (which I should probably call "allow list") means "it's ok for these to not be in a named scope" _and_ it will start off by coinciding with "the list of packages that are pending rename". And yes, once everything is renamed, the whitelist can be deleted. (I wouldn't actually be bothered with emptying it along the way... I'd just remove the whole thing at the end.)

**Comment by ambrinchaudhary** [comment]
Ok, then I guess is better to put this again as draft, to avoid accidentally forwards. 

Just a note @julien, some comments like in https://github.com/liferay-frontend/liferay-portal/pull/968#discussion_r613413763 may not worth to fix because I changed them in https://github.com/liferay-frontend/liferay-portal/pull/960. Maybe we should merge this two pulls in just one...

File: modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json
Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json:7]: @wincent sorry about that, I really need to remember this rule, I don't want you (or anyone else) to have to rename a gazillion packages again

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:37]: > And also, that looks like a global variable...
Yes, once more thing I didn't think about correctly

> what's going to happen if there are multiple editors on the page fighting to save/restore the "initial" class?
Probably something I don't want :joy:

> (Maybe this was discussed previously, so please forgive me if I missed that.)

It hasn't been discussed so thanks for the heads up. 
I'm thinking this should be in the component's "state" (i.e. `useState`)

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js
Comment (julien) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js:78]: To explain my intent (and give you back your gif): I thought of using a `Set` because I want each element to be unique, and now I'm thinking that, since I'm "adding" those items myself and that nothing from "outside" is going to touch `tags`, I could probably use a regular `Array`.

When it comes to iteration, you're right, that's something I didn't really think about.

The `path` object is a [`CKEDITOR.dom.elementPath`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_elementPath.html)(that's a link to the docs, but I guess you also want to have a look at the [source](https://github.com/ckeditor/ckeditor4/blob/eb2d11644a796cd13cd047cd55541430359f9317/core/dom/elementpath.js)), and the idea behind this iteration is the following:
If the `path` contains any of those "tags", then we should show this toolbar ... 

I actually changed [this part of the code](https://github.com/liferay-frontend/liferay-portal/blob/12dac565f1b763bf2966e5d391dadf03c05d341a/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BalloonEditor.js#L58-L82) locally, to be like this

```js
balloonToolbars.create({
    buttons: 'Bold,Italic,Underline,RemoveFormat,Link,NumberedList,BulletedList,JustifyLeft,JustifyCenter,JustifyRight,JustifyBlock,Anchor',
    cssSelector: '*',
});
```

(*It's not something I came up with, since it was already "like" this in the original "draft"; and unfortunately, I can't find that pull request, and I can't remember where it was ... but I think it doesn't really matter*)


The good news is that the result is exactly the same.  This also looks more "simple" even though I haven't analyzed what it's doing "under the hood", but it's probably better for now than my quadratic monster.

*Another thing I didn't say (and probably should have because it's not obvious), is that this will probably change. I'm still trying to figure out how to use this CKEditor plugin to behave like we want it to.*

File: modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json
Comment (wincent) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-sample-web/package.json]: The worry isn't about "gazillion" packages. Say Liferay adds "a few" packages (5, 10?) over the next few months. All we need is _one_ [bad actor](https://en.wikipedia.org/wiki/Nicolas_Cage) to come along and squat the package name on npm and we've got a potential headache. The odds of it happening are slim, and the scope of the likely damage if it does happen is not that great, but that's cold comfort because the badness of the worst case scenario is bad enough. `#bad`

bad

**Comment by ambrinchaudhary** [comment]
Just started reviewing :)

**Comment by julien** [comment]
I've had another thought about this PR (the good thing about these draft PRs is that we can actually take time to think about them) and I'm thinking we probably don't want and don't need the sample (`frontend-editor-ckeditor-sample-web`) portlet  for the following reasons:

1. It adds more code to maintain (for little value)
2. We're also planning to replace the current editor in `blogs-web` and page editor, so we'll already have at least more than one place to test this new editor

What do you think about that @ambrinchaudhary?
If you agree, I'll update the PR and remove the corresponding commits.

**Comment by ambrinchaudhary** [comment]
Hey @julien, 

for me this `frontend-editor-ckeditor-sample-web` portlet is very useful, it allows us to test this BalloonEditor properly in portal (you know it is not the same to be working on a demo that in portal). It does not add to much to maintain, because it only renders a taglib (`<liferay-editor:editor>`), and as we thought at the beginning, we can extend it to not just render the BalloonEditor, but any other editors available in portal so this could be also usefull for any developer or QAs. 

Also.... I don't think we can replace the Balloon Editor in Blogs or Page Editor until most of the work is done (which as you know won't be soon), so until then, if this portlet is not available there is no way to test the new editor in portal and face problems sooner.

**Comment by jbalsas** [comment]
Hey @julien, @ambrinchaudhary, one option to solve this would be to treat this sample portlet similar to what @zhangyan0701 and @john-co did with https://github.com/liferay-frontend/liferay-portal/tree/master/modules/apps/frontend-taglib/frontend-taglib-clay-test-alert-toast-sample-web

If we considered this a QA test portlet, then they could take it over and use it to build future coverage for our WYSIWYG editors.

**Comment by julien** [comment]
@jbalsas I actually didn't think about it but it probably make sense.
I'll check what needs to be changed to make that happen. @ambrinchaudhary, I guess you would also be happy about doing it this way, am I right?

**Comment by julien** [comment]
@ambrinchaudhary I removed the `.lfr-buildportal` file as I didn't see it [in the module QA is using to test toasts](https://github.com/liferay-frontend/liferay-portal/tree/master/modules/apps/frontend-taglib/frontend-taglib-clay-test-alert-toast-sample-web):bread:. I also rebased with `master` just to make sure everything was up to date.

**Comment by ambrinchaudhary** [comment]
LGTM @julien, thanks! I'm ok about sending this to master ;-)

**Comment by wincent** [comment]
> I'm ok about sending this to master

join-the-fun

`#AllTheCoolKidsAreDoingIt`

**Comment by ambrinchaudhary** [comment]
As work is done here and release is completed, going to forward this pull 🎉

**Comment by ambrinchaudhary** [comment]
ci:forward

**Pull request closed without merging**
